executor/write.go

```diff
@@ -82,8 +83,15 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 				return false, handleChanged, newHandle, 0, errors.Trace(err)
 			}
 		}
+
+		oldDataNull := oldData[i].IsNull()
+
+		if checkReplenish && oldDataAllNull && !oldDataNull {
+			oldDataAllNull = false
+		}
+
 		// Rebase auto increment id if the field is changed.
-		if mysql.HasAutoIncrementFlag(col.Flag) {
+		if mysql.HasAutoIncrementFlag(col.Flag) && !(checkReplenish && oldDataAllNull) {
```

Comment:
I think it is acceptable, but the root cause is not this. It seems MySQL sets the auto-id and checks NULL for it just before writing the row. It is much later than TiDB. Maybe someday we should refactor TiDB's implement.